plugin/evm/block.go

```diff
@@ -247,10 +268,33 @@ func (b *Block) verify(predicateContext *precompileconfig.PredicateContext, writ
 func (b *Block) verifyPredicates(predicateContext *precompileconfig.PredicateContext) error {
 	rules := b.vm.chainConfig.AvalancheRules(b.ethBlock.Number(), b.ethBlock.Timestamp())
 
+	switch {
+	case !rules.IsDUpgrade && rules.PredicatesExist():
+		return errors.New("cannot enable predicates before DUpgrade activation")
+	case !rules.IsDUpgrade:
+		return nil
+	}
+
+	predicateResults := results.NewPredicateResults()
 	for _, tx := range b.ethBlock.Transactions() {
-		if err := core.CheckPredicates(rules, predicateContext, tx); err != nil {
+		results, err := core.CheckPredicates(rules, predicateContext, tx)
+		if err != nil {
+			return err
+		}
+		predicateResults.SetTxPredicateResults(tx.Hash(), results)
+	}
+	// TODO: document required gas constraints to ensure marshalling predicate results does not error
+	predicateResultsBytes, err := predicateResults.Bytes()
+	if err != nil {
+		return fmt.Errorf("failed to marshal predicate results: %w", err)
+	}
+	headerPredicateResultsBytes := b.ethBlock.Extra()[params.DynamicFeeExtraDataSize:]
```

Comment:
could we add a comment to why this is safe?

Reply:
The length is checked in syntactic verification, but will add an extra check here to make it more obviously correct.